src/callbacks_step/analysis.jl

```diff
@@ -184,175 +188,208 @@ function (analysis_callback::AnalysisCallback)(integrator)
     mpi_println(" #elements:      " * @sprintf("% 14d", nelements(solver, cache)))
 
     # Level information (only show for AMR)
-    uses_amr = false
-    callbacks = integrator.opts.callback
-    if callbacks isa CallbackSet
-      for cb in callbacks.discrete_callbacks
-        if cb.affect! isa AMRCallback
-          uses_amr = true
-          break
-        end
-      end
-    end
-    if uses_amr
-      levels = Vector{Int}(undef, nelements(solver, cache))
-      min_level = typemax(Int)
-      max_level = typemin(Int)
-      for element in eachelement(solver, cache)
-        current_level = mesh.tree.levels[cache.elements.cell_ids[element]]
-        levels[element] = current_level
-        min_level = min(min_level, current_level)
-        max_level = max(max_level, current_level)
-      end
-
-      for level = max_level:-1:min_level+1
-        mpi_println(" ├── level $level:    " * @sprintf("% 14d", count(isequal(level), levels)))
-      end
-      mpi_println(" └── level $min_level:    " * @sprintf("% 14d", count(isequal(min_level), levels)))
-    end
+    print_amr_information(integrator.opts.callback, mesh, solver, cache)
     mpi_println()
 
     # Open file for appending and store time step and time information
     if mpi_isroot() && analysis_callback.save_analysis
       io = open(joinpath(analysis_callback.output_directory, analysis_callback.analysis_filename), "a")
-      mpi_isroot() && @printf(io, "% 9d", iter)
-      mpi_isroot() && @printf(io, "  %10.8e", t)
-      mpi_isroot() && @printf(io, "  %10.8e", dt)
-    end
-
-    # the time derivative can be unassigned before the first step is made
-    if t == integrator.sol.prob.tspan[1]
-      du_ode = similar(integrator.u)
+      @printf(io, "% 9d", iter)
+      @printf(io, "  %10.8e", t)
+      @printf(io, "  %10.8e", dt)
     else
-      du_ode = get_du(integrator)
+      io = devnull
     end
+
+    # Calculate current time derivative (needed for semidiscrete entropy time derivative, residual, etc.)
+    du_ode = first(get_tmp_cache(integrator))
     @notimeit timer() rhs!(du_ode, integrator.u, semi, t)
     GC.@preserve du_ode begin
       du = wrap_array(du_ode, mesh, equations, solver, cache)
+      l2_error, linf_error = analysis_callback(io, du, u, integrator.u, t, semi)
+    end # GC.@preserve du_ode
```

Comment:
```suggestion
    du = wrap_array(du_ode, mesh, equations, solver, cache)
    l2_error, linf_error = analysis_callback(io, du, u, integrator.u, t, semi)
```
The `GC.@preserve` should not be necessary anymore, since `du_ode` is not a reference to newly allocated memory but to a memory block that is already protected from being GC'd. Or am I missing something?

Reply:
Good catch, thanks :+1: Fixed in 2f447b5e9713b9b1aba6dfa886c3099bf323c887.